File: apps/api/plane/app/urls/permission.py
# SPDX-FileCopyrightText: 2023-present Plane Software, Inc.
# SPDX-License-Identifier: LicenseRef-Plane-Commercial
#
# Licensed under the Plane Commercial License (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
# https://plane.so/legals/eula
#
# DO NOT remove or modify this notice.
# NOTICE: Proprietary and confidential. Unauthorized use or distribution is prohibited.

from django.urls import path

from plane.app.views.permission import (
    ResourcePermissionEndpoint,
    RoleEndpoint,
    UserPermissionEndpoint,
    PermissionSchemeEndpoint,
    PermissionSchemeImpactEndpoint,
)


urlpatterns = [
    # Workspace-level resource permissions
    path(
        "workspaces/<str:slug>/resources/permissions/",
        ResourcePermissionEndpoint.as_view(),
        name="workspace-resource-permissions",
    ),
    # Role management endpoints
    path(
        "workspaces/<str:slug>/roles/",
        RoleEndpoint.as_view(),
        name="workspace-roles",
    ),
    path(
        "workspaces/<str:slug>/roles/<uuid:pk>/",
        RoleEndpoint.as_view(),
        name="workspace-role-detail",
    ),
    # User permission endpoints
    path(
        "workspaces/<str:slug>/permissions/",
        UserPermissionEndpoint.as_view(),
        name="workspace-user-permissions",
    ),
    path(
        "workspaces/<str:slug>/projects/<uuid:project_id>/permissions/",
        UserPermissionEndpoint.as_view(),
        name="project-user-permissions",
    ),
    path(
        "workspaces/<str:slug>/teamspaces/<uuid:team_space_id>/permissions/",
        UserPermissionEndpoint.as_view(),
        name="teamspace-user-permissions",
    ),
    # Permission scheme endpoints
    path(
        "workspaces/<str:slug>/permission-schemes/",
        PermissionSchemeEndpoint.as_view(),
        name="workspace-permission-schemes",
    ),
    path(
        "workspaces/<str:slug>/permission-schemes/<uuid:pk>/",
        PermissionSchemeEndpoint.as_view(),
        name="workspace-permission-scheme-detail",
    ),
    path(
        "workspaces/<str:slug>/permission-schemes/<uuid:pk>/impact/",
        PermissionSchemeImpactEndpoint.as_view(),
        name="workspace-permission-scheme-impact",
    ),
]

File: apps/web/core/store/favorite/permission.ts
/**
 * SPDX-FileCopyrightText: 2023-present Plane Software, Inc.
 * SPDX-License-Identifier: LicenseRef-Plane-Commercial
 *
 * Licensed under the Plane Commercial License (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * https://plane.so/legals/eula
 *
 * DO NOT remove or modify this notice.
 * NOTICE: Proprietary and confidential. Unauthorized use or distribution is prohibited.
 */

import { computedFn } from "mobx-utils";
// plane imports
import type { PermissionCheckArgs } from "@plane/types";

export interface FavoritePermissions {
  getCanView: (workspaceSlug: string) => boolean;
  getCanCreate: (workspaceSlug: string) => boolean;
  getCanEdit: (workspaceSlug: string, favoriteId: string) => boolean;
  getCanDelete: (workspaceSlug: string, favoriteId: string) => boolean;
}

type FavoritePermissionsArgs = {
  can: (args: PermissionCheckArgs) => boolean;
};

export class FavoritePermissionsInstance implements FavoritePermissions {
  constructor(private args: FavoritePermissionsArgs) {}

  getCanView: FavoritePermissions["getCanView"] = computedFn((workspaceSlug) =>
    this.args.can({ resource: "favorite", action: "view", workspaceSlug })
  );

  getCanCreate: FavoritePermissions["getCanCreate"] = computedFn((workspaceSlug) =>
    this.args.can({ resource: "favorite", action: "create", workspaceSlug })
  );

  getCanEdit: FavoritePermissions["getCanEdit"] = computedFn((workspaceSlug, favoriteId) =>
    this.args.can({
      resource: "favorite",
      action: "edit",
      workspaceSlug,
      resourceMeta: { resourceId: favoriteId },
    })
  );

  getCanDelete: FavoritePermissions["getCanDelete"] = computedFn((workspaceSlug, favoriteId) =>
    this.args.can({
      resource: "favorite",
      action: "delete",
      workspaceSlug,
      resourceMeta: { resourceId: favoriteId },
    })
  );
}
